metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/CheckpointingService.scala

```diff
@@ -0,0 +1,208 @@
+import io.iohk.metronome.hotstuff.consensus.basic.{Phase, QuorumCertificate}
+import io.iohk.metronome.hotstuff.service.ApplicationService
+import io.iohk.metronome.hotstuff.service.storage.{
+  BlockStorage,
+  ViewStateStorage
+}
+import io.iohk.metronome.storage.KVStoreRunner
+
+class CheckpointingService[F[_]: Sync: Applicative, N] private (
+    ledgerTree: Ref[F, Map[Block.Hash, (Ledger, Block.Header)]],
+    lastExecutedHeader: Ref[F, Block.Header],
+    ledgerUpdateSemaphore: Semaphore[F],
+    maxConcurrentUpdates: Int,
+    viewStateStorage: ViewStateStorage[N, CheckpointingAgreement],
+    ledgerStorage: LedgerStorage[N],
+    blockStorage: BlockStorage[N, CheckpointingAgreement]
+)(implicit storeRunner: KVStoreRunner[F, N])
+    extends ApplicationService[F, CheckpointingAgreement] {
+
+  override def createBlock(
+      highQC: QuorumCertificate[CheckpointingAgreement]
+  ): F[Option[Block]] = ???
+
+  override def validateBlock(block: Block): F[Option[Boolean]] = {
+    for {
+      lastExec <- lastExecutedHeader.get
+      ledger <- ledgerUpdateSemaphore.withPermit(
+        updateLedger(block, lastExec.height)
+      )
+      ledgerHashValid = block.header.postStateHash == ledger.hash
+      //TODO: Validate transactions PM-3131/3132
+      txsValid <- Some(false).pure[F]
+    } yield txsValid.map(_ && ledgerHashValid)
+  }
+
+  override def executeBlock(
+      block: Block,
+      commitQC: QuorumCertificate[CheckpointingAgreement],
+      commitPath: NonEmptyList[Block.Hash]
+  ): F[Boolean] = {
+    require(commitQC.phase == Phase.Commit, "Commit QC required")
+
+    if (commitQC.blockHash != block.hash) {
+      lastExecutedHeader.get
+        .flatMap { h =>
+          ledgerUpdateSemaphore.withPermit(
+            updateLedger(block, h.height)
+          )
+        }
+        .as(false)
+    } else {
+      Sync[F]
+        .bracket {
+          ledgerUpdateSemaphore.acquireN(maxConcurrentUpdates)
+        } { _ =>
+          for {
+            lastExec <- lastExecutedHeader.get
+            ledger   <- updateLedger(block, lastExec.height)
+            chkpOpt  <- constructCheckpoint(ledger, commitQC)
+            _        <- commitState(block.header, ledger)
+            -        <- chkpOpt.map(pushCheckpoint).getOrElse(().pure[F])
+          } yield false
+        } { _ =>
+          ledgerUpdateSemaphore.releaseN(maxConcurrentUpdates)
+        }
+    }
+  }
+
+  private def updateLedger(
+      block: Block,
+      executedHeight: Long
+  ): F[Ledger] = {
+
+    def loop(block: Block): F[Ledger] = {
+
+      getLedger(block.header.parentHash).flatMap {
+        case Some(ledger) =>
+          updateLedgerByBlock(ledger, block)
+
+        case None if block.header.height <= executedHeight =>
+          Sync[F].raiseError(
+            new IllegalStateException(
+              "Attempting to execute a branch that does not extend last executed block. Aborting execution"
+            )
+          )
```

Comment:
It is possible that by the time we try to evaluate the validity of a block, the execution went down a different path. Or that the jump synchronizer cleared out the ancestor block. 

In these cases you could just return `None`, as in you cannot produce the ledger that belongs to that block, which I reckon is only problematic during block execution. Validation and block creation could handle this issue without raising an exception. Or at least validation shouldn't throw an error.